addons/product/models/product_attribute.py

```diff
@@ -447,7 +447,15 @@ def write(self, values):
                         _("You cannot change the product of the value %s set on product %s.") %
                         (ptav.display_name, ptav.product_tmpl_id.display_name)
                     )
-        return super(ProductTemplateAttributeValue, self).write(values)
+        res = super(ProductTemplateAttributeValue, self).write(values)
+
+        if 'exclude_for' in values:
+            self.product_tmpl_id._create_variant_ids()
+            # prevent change if exclusion deleted template by deleting last variant
+            if self.exists() != self:
+                raise UserError("Forbidden action. You cannot exclude the last variant in this context. Please archive or delete your product directly if intended.")
```

Comment:
Could the check be moved in `_create_variant_ids` instead?
Is there any scenario where we call it and expect the template to be deleted?

Reply:
If I move the check in `_create_variant_ids` there might be an impact for such scenario. By doing this way, at the root of the new code, we mitigate this risk.

Reply:
I know it would have an impact, but I think it should.

The unlink variant thing is used when manually deleting a variant, it should probably not be allowed when generating them automatically from any situation.

Reply:
I see what you mean. It is not a behavior that we want in any case. I moved it into `_create_variant_ids`